File: Rendering/GridRenderer.cs
using MCBp.Canvas;

namespace MCBp.Rendering;

public static class GridRenderer
{
    private static readonly Pen GridPen = new(Color.FromArgb(40, 180, 180, 180), 1f);
    private static readonly Pen ChunkPen = new(Color.FromArgb(80, 120, 120, 120), 1f);
    private static readonly Pen OriginPen = new(Color.FromArgb(150, 80, 80, 80), 1.5f);
    private static readonly Pen BoundaryPen = new(Color.FromArgb(120, 100, 100, 100), 2f);

    public static void Draw(Graphics g, Camera camera, Size clientSize, bool showChunks)
    {
        var visRect = camera.GetVisibleGridRect(clientSize);
        int cs = camera.CellSize;

        if (cs >= 4)
        {
            for (int gx = visRect.Left; gx <= visRect.Right; gx++)
            {
                float sx = gx * cs + camera.OffsetX;
                bool isChunk = showChunks && gx % 16 == 0;
                var pen = isChunk ? ChunkPen : GridPen;
                g.DrawLine(pen, sx, 0, sx, clientSize.Height);
            }

            for (int gy = visRect.Top; gy <= visRect.Bottom; gy++)
            {
                float sy = gy * cs + camera.OffsetY;
                bool isChunk = showChunks && gy % 16 == 0;
                var pen = isChunk ? ChunkPen : GridPen;
                g.DrawLine(pen, 0, sy, clientSize.Width, sy);
            }
        }

        float y0 = camera.OffsetY;
        if (y0 >= 0 && y0 <= clientSize.Height)
            g.DrawLine(OriginPen, 0, y0, clientSize.Width, y0);

        float x0 = camera.OffsetX;
        if (x0 >= 0 && x0 <= clientSize.Width)
            g.DrawLine(OriginPen, x0, 0, x0, clientSize.Height);

        float yMin = Camera.MinY * cs + camera.OffsetY;
        float yMax = (Camera.MaxY + 1) * cs + camera.OffsetY;

        if (yMin >= -2 && yMin <= clientSize.Height + 2)
            g.DrawLine(BoundaryPen, 0, yMin, clientSize.Width, yMin);
        if (yMax >= -2 && yMax <= clientSize.Height + 2)
            g.DrawLine(BoundaryPen, 0, yMax, clientSize.Width, yMax);
    }
}

File: src/core/renderers/gridRenderer.ts
import { Camera } from '../camera';
import { Size } from '../types';

export function drawGrid(
  ctx: CanvasRenderingContext2D,
  camera: Camera,
  clientSize: Size,
  showChunks: boolean
): void {
  const visRect = camera.getVisibleGridRect(clientSize);
  const cs = camera.cellSize;

  if (cs >= 4) {
    ctx.lineWidth = 1;

    for (let gx = visRect.left; gx <= visRect.right; gx++) {
      const sx = gx * cs + camera.offsetX;
      const isChunk = showChunks && gx % 16 === 0;

      ctx.strokeStyle = isChunk
        ? 'rgba(160, 160, 160, 0.45)'
        : 'rgba(210, 210, 210, 0.25)';
      ctx.beginPath();
      ctx.moveTo(sx, 0);
      ctx.lineTo(sx, clientSize.height);
      ctx.stroke();
    }

    for (let gy = visRect.top; gy <= visRect.bottom; gy++) {
      const sy = gy * cs + camera.offsetY;
      const isChunk = showChunks && gy % 16 === 0;

      ctx.strokeStyle = isChunk
        ? 'rgba(160, 160, 160, 0.45)'
        : 'rgba(210, 210, 210, 0.25)';
      ctx.beginPath();
      ctx.moveTo(0, sy);
      ctx.lineTo(clientSize.width, sy);
      ctx.stroke();
    }
  }

  const y0 = camera.offsetY;
  if (y0 >= 0 && y0 <= clientSize.height) {
    ctx.strokeStyle = 'rgba(80, 80, 80, 0.7)';
    ctx.lineWidth = 1.5;
    ctx.beginPath();
    ctx.moveTo(0, y0);
    ctx.lineTo(clientSize.width, y0);
    ctx.stroke();
  }

  const x0 = camera.offsetX;
  if (x0 >= 0 && x0 <= clientSize.width) {
    ctx.strokeStyle = 'rgba(80, 80, 80, 0.7)';
    ctx.lineWidth = 1.5;
    ctx.beginPath();
    ctx.moveTo(x0, 0);
    ctx.lineTo(x0, clientSize.height);
    ctx.stroke();
  }

  const yMin = Camera.MinY * cs + camera.offsetY;
  const yMax = (Camera.MaxY + 1) * cs + camera.offsetY;

  ctx.strokeStyle = 'rgba(120, 120, 120, 0.6)';
  ctx.lineWidth = 2;

  if (yMin >= -2 && yMin <= clientSize.height + 2) {
    ctx.beginPath();
    ctx.moveTo(0, yMin);
    ctx.lineTo(clientSize.width, yMin);
    ctx.stroke();
  }

  if (yMax >= -2 && yMax <= clientSize.height + 2) {
    ctx.beginPath();
    ctx.moveTo(0, yMax);
    ctx.lineTo(clientSize.width, yMax);
    ctx.stroke();
  }
}
